SARIX and GBQR take NSSP Data

lshandross opened the pull request:
Updates the SARIX and GBQR model classes to take NSSP as a data source, as well as the tests/test data to test the new functionality. Likely most easily reviewed by looking at individual commits, and I could also walk someone through it if need be

lshandross committed:
Update `iddata` requirement

lshandross committed:
Update sarix to handle nssp data

lshandross committed:
Update gbqr to handle nssp data

lshandross committed:
Restrict NSSP GBQR preds to [0, 1]

lshandross committed:
Update tests and test data for NSSP support

* Add tests (and test data) for SARIX and GBQR for NSSP data source
* Rename tests and test data to specify the main data source (i.e. NHSN or NSSP)
* Update file paths in tests to be dynamic (not hard-coded)

lshandross committed:
Update requirements text files

nickreich commented on requirements/requirements-dev.txt:
```diff
@@ -147,7 +147,7 @@ ruff==0.7.2
 s3fs==2024.10.0
     # via iddata
-sarix @ git+https://github.com/elray1/sarix@1c8995942d49afbb66637f0dc0662e1248606af4
+sarix @ git+https://github.com/elray1/sarix@c8921734a4328a372cf23ff343e781bc08dd4d36
```
small suggestion (not part of a larger review yet) but I think we are moving towards hosting our own version of sarix here: https://github.com/reichlab/sarix/tree/main and not pointing at Evan's repo anymore...

nickreich commented on requirements/requirements.txt:
```diff
@@ -117,7 +117,7 @@ rich==13.9.4
 s3fs==2024.10.0
     # via iddata
-sarix @ git+https://github.com/elray1/sarix@1c8995942d49afbb66637f0dc0662e1248606af4
+sarix @ git+https://github.com/elray1/sarix@c8921734a4328a372cf23ff343e781bc08dd4d36
```
same as above

lshandross committed:
Use reichlab/sarix as dependency

matthewcornell commented on src/idmodels/sarix.py:
```diff
@@ -13,10 +13,27 @@ def __init__(self, model_config):
 
     def run(self, run_config):
+        valid_sources = np.array(["nhsn", "nssp"])
+        if ~np.isin(np.array(self.model_config.sources), valid_sources).all():
```
I think we should use `not` (logical operation) instead of `~` (bitwise operation). BUT: I might be wrong:

> The exception is when working with specialized libraries like Pandas or NumPy, where ~ is overloaded to perform logical negation on certain data structures

trobacker commented on src/idmodels/sarix.py:
```diff
@@ -13,10 +13,27 @@ def __init__(self, model_config):
+            df = fdl.load_data(nssp_kwargs={"as_of": None, "disease": run_config.disease},
+                               sources=self.model_config.sources,
+                               power_transform=self.model_config.power_transform)
+            df["source"].unique()
```
Can remove this line (L34)

trobacker commented on src/idmodels/gbqr.py:
```diff
@@ -153,7 +170,12 @@ def _train_gbq_and_predict(self, run_config,
+            target_name = "wk inc " + run_config.disease + " hosp"
+        if "nssp" in preds_df["source"].unique():
+            target_name = "wk inc " + run_config.disease + " prop ed visits"
+
```
Can make this an else-if for clarity

lshandross committed:
Remove debugging code

lshandross committed:
Use if-elif instead of two ifs

lshandross committed:
Use not instead of ~

lshandross committed:
Update sarix package dependency version

lshandross committed:
Update sarix dependency commit hash

lshandross committed:
Convert NSSP observed percentage to proportion

lshandross committed:
Update SARIX NSSP integration test data

lshandross committed:
Update sarix for hsa and state-level forecasting

lshandross committed:
Refactor sarix tests to create configs w helpers

lshandross committed:
Add more sarix tests + update test data

lshandross committed:
Update + refactor gbqr tests

* Refactor gbqr tests to create configs with helpers
* Add more gbqr tests + update test data

lshandross committed:
Update gbqr for hsa and state-level forecasting

lshandross committed:
Restrict rates to counts conversion to nhsn; update tests + test data

lshandross committed:
Update `iddata` dependency in requirements files

lshandross committed:
Merge branch 'main' into ls/sarix-gbqr-takes-nssp-data/9

lshandross committed:
Fix linter I001 error

lshandross committed:
Update sarix fourier tests + config creation methods

* Update sarix fourier tests to use separate states and hsas fields, same logic as other tests, config creation helpers when applicable
* Add args to config creation helper methods to be more flexible

lshandross committed:
Update nssp tests to use `ref_date=2025-09-20` + update test data

lshandross committed:
Internal nssp loading uses provided ref date

matthewcornell reviewed (dismissed):
I didn't do a deep dive into the logic, but it looks good from a quick-ish scan. Exciting!

nickreich commented on pyproject.toml:
```diff
@@ -15,7 +15,7 @@ dependencies = [
     "numpy",
     "pandas",
-    "sarix @ git+https://github.com/reichlab/sarix@35eea2379a9790e0457b1aed41d13509e5d5056f",
+    "sarix @ git+https://github.com/reichlab/sarix",
```
would it be worth pinning a specific commit here, as we have done in previous versions? I think hubverse dev standards says this is good practice.

nickreich replied:
Noting that also the specific hash that was deleted is still the pinned requirement in requirements.txt, so either it could be kept as is, or perhaps the requirements files need to be regenerated again?

lshandross replied:
There's no need to pin a specific commit in the pyproject file because this information gets recorded in the requirements file(s), and from what I understand it's actually recommended not to pin in this file. The requirements files don't automatically update the versions; you'll have to run a specific command to update the commit hash. So I think this is fine as-is

nickreich commented on src/idmodels/gbqr.py:
```diff
@@ -31,15 +31,38 @@ def run(self, run_config):
-        if run_config.locations is not None:
-            df = df.loc[df["location"].isin(run_config.locations)]
+        if "nhsn" in self.model_config.sources:
+            df = fdl.load_data(nhsn_kwargs={"as_of": run_config.ref_date, "disease": run_config.disease},
+                               ilinet_kwargs=ilinet_kwargs,
+                               flusurvnet_kwargs=flusurvnet_kwargs,
```
This is a question more than a comment on the code: is the implication here that the ilinet and flusurvnet data are treated more as accessory data sources to the primary data source of nssp or nhsn? Meaning that nhsn or nssp data are loaded (only one) and then if flusurvnet and/or ilinet are specified, they are loaded as "accessory/covariate" sources for the primary source?

lshandross replied:
Yes, ilinet and flusurvnet are used to augment the nhsn data (and if we find it useful could also be used to do the same for nssp), but we never use them to forecast on their own.

nickreich commented on src/idmodels/gbqr.py:
```diff
@@ -31,15 +31,38 @@ def run(self, run_config):
+            raise ValueError("User must request a non-empty set of locations to forecast for.")
+
+        if (run_config.states != []) & (run_config.hsas != []):
+            raise NotImplementedError("Functionality for simultaneously forecasting state- and hsa-level locations is not yet implemented.")
```
if we haven't already, let's file an issue for this.

lshandross replied:
Filed #17!

nickreich commented on src/idmodels/gbqr.py:
```diff
@@ -149,11 +172,20 @@ def _train_gbq_and_predict(self, run_config,
+            target_name = "wk inc " + run_config.disease + " prop ed visits"
+
+        preds_df["value"] = np.maximum(preds_df["value"], 0.0)
+        preds_df = self._format_as_flusight_output(preds_df, run_config.ref_date, target_name)
```
I'm noting that this method has different arguments than the old line 156 that it replaced. is that expected and ok?

nickreich replied:
does it require updates to _format_as_flusight_output?

lshandross replied:
Yes, I replaced an argument in the function so that `disease` got superseded by `target_name`. Previously, `disease` was used to construct the target name, but I moved the construction elsewhere to avoid additional if statements

nickreich commented on src/idmodels/sarix.py:
```diff
@@ -34,8 +57,12 @@ def run(self, run_config):
-
+   
         xy_colnames = self.model_config.x + ["inc_trans_cs"]
+        df = df.query("wk_end_date >= '2022-10-01'").interpolate()
```
if there is some data interpolation going on here, could we add a comment to that effect and possible throw a message to the user?

lshandross replied:
Besides a comment, is a general warning that missing values will be interpolated sufficient?

lshandross replied:
Filed #18

nickreich reviewed (dismissed):
My comments are all minor and should not stop a merge for expediency. If we don't address one of my comments, could we file an issue for it?

lshandross committed:
Remove redundant code

lshandross committed:
Update unique location calculation

Better match original code + Avoid error whenNot all requested locations in data

lshandross committed:
Add in-line comment about interpolation

nickreich reviewed (dismissed):
reviewed synchronously

lshandross committed:
Increment to v1.0.0

New functionality added: allow NSSP-based forecasting
Breaking change: run_config.locations superseded by run_config.states and run_config.hsas

nickreich reviewed (approved):
LGTM! thanks for the iteration.
